Support for Copy on Write semantics (reflink)

mwanner opened the pull request:
These two commits add support for Copy on Write semantics to `fs_extra` via the `reflink` crate.

webdesus commented:
HI @mwanner, nice job :blush: 

But I, like many programmers for minimalism in the dependency tree. Can you do attach `reflinks` by option([features](https://doc.rust-lang.org/cargo/reference/features.html)) for cargo file? Like [Uuid](https://crates.io/crates/uuid) crate attach `serde` crate. 

I would really appreciate foryour help :blush:

mwanner commented:
@webdesus It's been an awful while, but... I finally updated this PR to make all reflink specific stuff conditional on the non-default feature `refilnk`. Is that what you've been asking for?

mwanner committed:
Deduplicate by implementing From<dir::CopyOptions> for file::CopyOptions.

mwanner committed:
Add the option to use reflinks (i.e. copy on write semantics).

Using the reflink crate, with the option to disallow completely
(never), auto (use when available), or always (fail if not possible).
Fully conditional on a new feature "reflink".

mwanner commented:
@webdesus Anything else you need here?

mwanner commented:
> But I, like many programmers for minimalism in the dependency tree. Can you do attach `reflinks` by option([features](https://doc.rust-lang.org/cargo/reference/features.html)) for cargo file? Like [Uuid](https://crates.io/crates/uuid) crate attach `serde` crate.

Hey @webdesus , this PR is updated to support `reflinks` only when specifically enabled through a corresponding feature.

Mind taking another look at the PR, please?
